### ISSUE TITLE
axis_angle_to_rotation_matrix returns non-orthogonal matrices: det = 0.9999974535249636 in float64 at every angle, because eps is added to theta in the axis normalization

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#3947]
`axis_angle_to_rotation_matrix` adds `eps = 1e-6` to `theta` in the axis normalization, so the returned matrix is not a rotation matrix: in **float64** at `theta = pi/2` it has `det = 0.9999974535249636` and `max|R Rᵀ − I| = 2.5464750363912714e-06`. The error is present at every angle tested, does not shrink with dtype, and is six orders of magnitude larger than float64 precision. Below `theta = 1e-3` a second, undocumented branch returns the raw first-order Taylor matrix `[[1, −rz, ry], [rz, 1, −rx], [−ry, rx, 1]]`, whose determinant is `1 + theta²` — also not a rotation. kornia's own quaternion route computes the same rotation exactly, so no third-party reference is needed to demonstrate the defect.

The source (`kornia/geometry/conversions.py:433-435`, and the branch at `:469-497`):

```python
def _compute_rotation_matrix(axis_angle: torch.Tensor, theta2: torch.Tensor, eps: float = 1e-6) -> torch.Tensor:
    theta = torch.sqrt(theta2.clamp(min=1e-12))  # clamping to ensure no nan gradients
    wxyz = axis_angle / (theta.unsqueeze(-1) + eps)  # (B, 3)
```

The unit axis is computed as `v / (theta + eps)` instead of `v / theta`, which shrinks it by a factor `theta / (theta + eps) = 1 − eps/theta + …`. Rodrigues' formula then builds a matrix from a *non-unit* axis, which is not orthogonal. The `theta2.clamp(min=1e-12)` on the line above already handles the NaN-gradient concern that `eps` looks like it is there for (see #2924, closed) — the `+ eps` on the next line is doing nothing for the gradient and everything to the value.

## Repro

```python
import torch
from kornia.geometry.conversions import (
    axis_angle_to_quaternion as aa2q,
    axis_angle_to_rotation_matrix as aa2R,
    quaternion_to_rotation_matrix as q2R,
)

I3 = torch.eye(3, dtype=torch.float64)
v = torch.tensor([[0.0, 0.0, torch.pi / 2]], dtype=torch.float64)

R = aa2R(v)[0]
print("aa2R : det =", torch.linalg.det(R).item(), " max|RR^T-I| =", (R @ R.T - I3).abs().max().item())

# kornia's own reference for the *same* rotation, via the quaternion route
Rq = q2R(aa2q(v))[0]
print("q2R  : det =", torch.linalg.det(Rq).item(), " max|RR^T-I| =", (Rq @ Rq.T - I3).abs().max().item())
print("max|aa2R(v) - q2R(aa2q(v))| =", (R - Rq).abs().max().item())

print("\n# det(aa2R) over the whole angle range, float64")
for theta in [1e-3, 1e-2, 0.1, 1.0, torch.pi / 2, torch.pi - 1e-6, torch.pi]:
    Rt = aa2R(torch.tensor([[0.0, 0.0, theta]], dtype=torch.float64))[0]
    print(f"  theta={theta:<20.15g} det={torch.linalg.det(Rt).item():.16f}")

print("\n# the Taylor branch below theta = 1e-3 (theta**2 > 1e-6), float64")
for theta in [9.9e-4, 1.0e-3, 1.001e-3]:
    Rt = aa2R(torch.tensor([[0.0, 0.0, theta]], dtype=torch.float64))[0]
    print(f"  theta={theta:<10} det={torch.linalg.det(Rt).item():.16f}")

print("\n# the same defect in float32 (identical magnitude)")
R32 = aa2R(torch.tensor([[0.0, 0.0, torch.pi / 2]], dtype=torch.float32))[0]
print("  float32 det =", torch.linalg.det(R32).item())
```

## Observed

```
aa2R : det = 0.9999974535249636  max|RR^T-I| = 2.5464750363912714e-06
q2R  : det = 1.0  max|RR^T-I| = 0.0
max|aa2R(v) - q2R(aa2q(v))| = 1.273238328769466e-06

# det(aa2R) over the whole angle range, float64
  theta=0.001                det=1.0000009999999999
  theta=0.01                 det=0.9999999700052495
  theta=0.1                  det=0.9999997007535928
  theta=1                    det=0.9999976644625806
  theta=1.5707963267949      det=0.9999974535249636
  theta=3.14159165358979     det=0.9999987267606578
  theta=3.14159265358979     det=0.9999987267610633

# the Taylor branch below theta = 1e-3 (theta**2 > 1e-6), float64
  theta=0.00099    det=1.0000009801000000
  theta=0.001      det=1.0000009999999999
  theta=0.001001   det=0.9999999970014948

# the same defect in float32 (identical magnitude)
  float32 det = 0.9999975562095642
```

## Expected

`axis_angle_to_rotation_matrix` should return an element of SO(3): `det = 1` and `R Rᵀ = I` to machine precision in the working dtype. `q2R(aa2q(v))` in the same file, on the same input, achieves exactly that (`det = 1.0`, `max|R Rᵀ − I| = 0.0` in float64), and the two paths disagree by `1.273238328769466e-06` — so this is not an inherent limit of the operation, it is this implementation.

Two separate consequences a user hits:

1. **Every downstream consumer inherits a `~1e-6` error even in float64.** The `rotation_matrix_to_axis_angle ∘ axis_angle_to_rotation_matrix` round-trip is accurate only to `~1e-6` in float64 for a generic axis (measured: `8.0e-07` at `theta = 1e-3`, `5.4e-07` at `theta = pi`) — the loss is entirely on the `aa2R` leg, since `axis_angle_to_quaternion ↔ quaternion_to_axis_angle` round-trips exactly (`<= 2e-19` in float64). Anyone using `aa2R` as ground truth in a test, or composing several of these matrices, accumulates a bias that no dtype increase removes.
2. **Two different matrices for the same rotation.** `aa2R(v)` and `q2R(aa2q(v))` are both "the" rotation matrix for `v` but differ at `1.3e-6`, which makes the two routes non-interchangeable and makes any cross-check between them fail at a tolerance that looks like a float32 tolerance while running in float64.

The fix is to normalize by `theta` and put the singularity guard where it belongs — in the gradient path only, e.g. a double-`where` on `theta2` so the value uses the exact `theta` while the backward pass never differentiates through `sqrt(0)` (the `theta2.clamp(min=1e-12)` already present is the same idea applied one line earlier). Reference: Rodrigues' formula as in `cv::Rodrigues` (OpenCV), which normalizes by the exact angle and special-cases only `theta == 0`.

The `theta² > 1e-6` Taylor branch needs a decision in the same change: today it is an undocumented threshold that returns `I + [v]ₓ` verbatim (`det = 1 + theta²`, so `1.0000009999999999` at `theta = 1e-3` — *worse* than the main branch at that angle). A second-order Taylor expansion, or simply routing everything through the guarded Rodrigues form, removes the discontinuity at the threshold. Note the two branches disagree in det direction across the boundary (`1.0000009999999999` at `1e-3` vs `0.9999999970014948` at `1.001e-3`).

Related closed prior art: #317 (round-trip inconsistency through `angle_axis_to_rotation_matrix`, closed but the eps is still there) and #2924 (the NaN backward that motivated the guards). This is the same defect family as #3939 (`cart2pol`'s `sqrt(x² + y² + eps)`) — an epsilon folded into a returned value instead of into the gradient path — but a different function with a different magnitude, so it is filed separately.

## Classification

**Broken output.** `det != 1` is wrong under every convention; there is no reading of "axis-angle to rotation matrix" under which a non-orthogonal matrix is the intended answer, and kornia's own alternative path proves the exact answer is reachable. Per the reference-core lifecycle (Part I / Part III §2) that would allow it to land as an ordinary bugfix rather than waiting for the coordinated Repair & Deprecation window.

The caveat worth a maintainer call: the fix changes the default output of **every** `axis_angle_to_rotation_matrix` call by `~1e-6`, so it cannot land under byte-identity discipline, and any existing test that pinned `aa2R` output at a `1e-7`-ish tolerance will need re-baselining (including this file's own second doctest, whose `1.0000e+00` display currently hides the defect — the true value is `0.9999987483024597`).

## Environment

kornia 0.9.0rc1 @ `3358ef16`, torch 2.9.1, cpu. Executed from a worktree carrying the batch-3a docs/tests commits on top of `3358ef16`; `axis_angle_to_rotation_matrix` is byte-identical to `origin/main` there (no diff hunk lands in it; function body compares AST-identical).

Posted on behalf of @ducha-aiki by Claude (Fable 5).

**Comment by ducha-aiki** [comment]
Scope correction to this issue's headline number, found while writing the convention pins for it (batch 3b of #3941). The defect is real and the diagnosis (an `eps` added to `theta` in the axis normalisation) is unchanged — but **the determinant is not `0.9999974535249636` "at every angle"**. It varies with `theta`, and the low-angle branch errs in the opposite direction:

```python
import math, torch
from kornia.geometry.conversions import axis_angle_to_rotation_matrix as aa2R

for theta in (0.001, 0.01, 0.1, 0.5, math.pi / 2, 2.0, math.pi):
    v = torch.tensor([[0.0, 0.0, 1.0]], dtype=torch.float64) * theta
    print(theta, float(torch.linalg.det(aa2R(v)[0])))
```

```
theta        det (float64)
0.001        1.0000009999999999     <- first-order Taylor branch, det > 1
0.01         0.9999999700052495
0.1          0.9999997007535928
0.5          0.9999985909395367
pi/2         0.9999974535249636     <- the number in the title
2.0          0.9999977570342060
pi           0.9999987267610633
```

So `0.9999974535249636` is the worst case, reached near `theta = pi/2`, not a constant. Two related scopings:

- **It is `theta`-dependent but essentially axis-independent.** At `theta = pi/2` the determinant agrees to ~3e-16 across axes (`[0,0,1]` and `[1,0,0]` → `...9636`, `[1,1,0]` → `...9634`, `[0.3,-0.9,0.1]` → `...9637`). The orthogonality residual `max|R Rᵀ − I|` *is* axis-dependent: `2.5464750363912714e-06` about `+z`, `2.091747640764474e-06` for a generic axis.
- **The `theta <= 1e-3` first-order branch is a floating-point knife-edge, not a clean threshold.** The test is on `theta² > eps`, so `[0, 0, 1e-3]` in float64 computes `theta² == 1e-06` exactly and takes the Taylor branch (`det = 1.000001`), while a generic axis scaled to the same `theta` computes `1.0000000000000002e-06` and takes the general branch (`det = 0.9999999970044947`). Float32 lands differently again. An input's branch depends on its axis and dtype, so no crisp angle threshold can be quoted.

None of this changes whether it should be fixed — a rotation conversion returning a non-orthogonal matrix is a bug at any magnitude, and the fix (don't perturb `theta`; guard the division instead, as `q2R(aa2q(v))` effectively does — it gives `det = 1.0` and `max|R Rᵀ − I| = 2.2e-16`) addresses every row above at once.

The batch-3b PR's convention pins and `.. warning::` blocks state the scoped facts, so the docs and this issue now agree.

Posted on behalf of @ducha-aiki by Claude (Fable 5).
